### PR TITLE
[GDScript] Prevent running `String` number functions on invalid literal

### LINKED ISSUE #83521: Invalid (or incomplete) hex number "0x" floods error messages
### Godot version

v4.2.beta1.official [b1371806a]

### System information

Godot v4.2.beta1 - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1060 6GB (NVIDIA; 31.0.15.3640) - Intel(R) Core(TM) i5-3570K CPU @ 3.40GHz (4 Threads)

### Issue description

When I start writing a hex number in GDScript code, parser starts flooding error messages when the hex number is not yet valid.

First a single error appears, but very soon there will be 2 or 3 more. Every save (Ctrl-S) adds 5 more error messages.

I would expect to get only one error message.

kuva

### Steps to reproduce

Write `var i = 0` first and then add `x`. 

```gdscript
extends Node2D

func _ready() -> void:
	pass
	var i = 0x
```

Output:

```
  Invalid hexadecimal notation character "x" (U+0078) in string "0x".
  Invalid hexadecimal notation character "x" (U+0078) in string "0x".
  ...
```

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by AThousandShips**
Prevents printing excessive errors.

We could also just remove the error print in `hex_to_int` but it was added quite a while ago and is a good indication of what's wrong when parsing strings IMO, and at least I don't think it should be removed just to prevent an annoying print in the GDScript parser (I'd even argue we might want to *add* errors to `bin_to_int` and the others)

Literals containing decimal points will already stop consuming input before them so that won't trigger an error print on that part (other than the GDScript tokenizer error, but the annoying part is the print to console)

* Fixes: https://github.com/godotengine/godot/issues/83521

### CODE REVIEW

File: modules/gdscript/gdscript_tokenizer.cpp [hunk 1]
```diff
@@ -820,6 +824,19 @@ GDScriptTokenizer::Token GDScriptTokenizer::number() {
 		}
 	}
 
+	if (need_digits) {
+		// No digits in hex or bin literal.
+		Token error = make_error(vformat(R"(Expected %s digit after "0%c".)", (base == 16 ? "hexadecimal" : "binary"), (base == 16 ? 'x' : 'b')));
+		error.start_column = column;
+		error.leftmost_column = column;
+		error.end_column = column + 1;
+		error.rightmost_column = column + 1;
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_tokenizer.cpp:833]: Unsure if the full error info is required here, but added it for completeness, unsure what the difference here is to not using it like below

File: modules/gdscript/gdscript_tokenizer.cpp [hunk 2]
```diff
@@ -820,6 +824,19 @@ GDScriptTokenizer::Token GDScriptTokenizer::number() {
...
+		push_error(error);
+
+		// Return empty literal.
+		return make_literal((int64_t)0);
```
Comment (dalexeev) [inline modules/gdscript/gdscript_tokenizer.cpp:837]:
```suggestion
		return error;
```

Comment (AThousandShips) [reply modules/gdscript/gdscript_tokenizer.cpp:837]: Will do, it prints different errors about expecting an expression but makes sense to error out

Could return errors more here to keep it unified, what do you think?

Comment (dalexeev) [reply modules/gdscript/gdscript_tokenizer.cpp]: I think `push_error()` is used when we want to continue reading the token despite an error (for example inside string literals) and `return make_error()`/`return error` when the current token has most likely ended.

Comment (AThousandShips) [reply modules/gdscript/gdscript_tokenizer.cpp]: Could swap for the others then too if desired

File: modules/gdscript/gdscript_tokenizer.cpp (hunk 1 above)
Comment (dalexeev) [reply modules/gdscript/gdscript_tokenizer.cpp:833]: I think it is right. Such errors should be reported as a cursor rather than a selection.

**Review by dalexeev — APPROVED** [review]
Looks good to me. `push_error()` vs `return error` can be improved later (like lines 844-845 and 848).

**Comment by akien-mga** [comment]
Thanks!

**Comment by AThousandShips** [comment]
Thank you!

**Comment by akien-mga** [comment]
Cherry-picked for 4.2.2.

**Pull request merged**